gusto/time_discretisation/implicit_runge_kutta.py

```diff
@@ -91,31 +97,108 @@ def setup(self, equation, apply_bcs=True, *active_labels):
 
         super().setup(equation, apply_bcs, *active_labels)
 
-        self.k = [Function(self.fs) for i in range(self.nStages)]
+        if self.rk_formulation == RungeKuttaFormulation.predictor:
+            self.xs = [Function(self.fs) for _ in range(self.nStages)]
+        elif self.rk_formulation == RungeKuttaFormulation.increment:
+            self.k = [Function(self.fs) for _ in range(self.nStages)]
+        elif self.rk_formulation == RungeKuttaFormulation.linear:
+            raise NotImplementedError(
+                'Linear Implicit Runge-Kutta formulation is not implemented'
+            )
+        else:
+            raise NotImplementedError(
+                'Runge-Kutta formulation is not implemented'
+            )
 
     def lhs(self):
         return super().lhs
 
     def rhs(self):
         return super().rhs
 
-    def solver(self, stage):
-        residual = self.residual.label_map(
-            lambda t: t.has_label(time_derivative),
-            map_if_true=drop,
-            map_if_false=replace_subject(self.xnph, self.idx),
-        )
+    def res(self, stage):
+        """Set up the discretisation's residual for a given stage."""
+        # Add time derivative terms  y_s - y^n for stage s
         mass_form = self.residual.label_map(
             lambda t: t.has_label(time_derivative),
             map_if_false=drop)
-        residual += mass_form.label_map(all_terms,
-                                        replace_subject(self.x_out, self.idx))
+        residual = mass_form.label_map(all_terms,
+                                       map_if_true=replace_subject(self.x_out, old_idx=self.idx))
+        residual -= mass_form.label_map(all_terms,
+                                        map_if_true=replace_subject(self.x1, old_idx=self.idx))
+        # Loop through stages up to s-1 and calcualte/sum
+        # dt*(a_s1*F(y_1) + a_s2*F(y_2)+ ... + a_{s,s-1}*F(y_{s-1}))
+        for i in range(stage):
+            r_imp = self.residual.label_map(
+                lambda t: not t.has_label(time_derivative),
+                map_if_true=replace_subject(self.xs[i], old_idx=self.idx),
+                map_if_false=drop)
+            r_imp = r_imp.label_map(
+                all_terms,
+                map_if_true=lambda t: Constant(self.butcher_matrix[stage, i])*self.dt*t)
+            residual += r_imp
+        # Calculate and add on dt*a_ss*F(y_s)
+        r_imp = self.residual.label_map(
+            lambda t: not t.has_label(time_derivative),
+            map_if_true=replace_subject(self.x_out, old_idx=self.idx),
+            map_if_false=drop)
+        r_imp = r_imp.label_map(
+            all_terms,
+            map_if_true=lambda t: Constant(self.butcher_matrix[stage, stage])*self.dt*t)
+        residual += r_imp
+        return residual.form
+
+    @property
+    def final_res(self):
+        """Set up the discretisation's final residual."""
+        # Add time derivative terms  y^{n+1} - y^n
```

Comment:
Can we make clear that this is only for the `predictor` formulation?

Reply:
Done